[#279] Processor to create (new) PyBossa tasks for new or modified indications

victor-nitu opened the pull request:
This new processor makes a query for new or modified indications, extracted from FDA approvals. The result is used to create new PyBossa tasks, deduplicate and `POST` them to the server as new tasks.

The main issue here is with API rate limiting, so there are a few mechanisms to avoid data loss due to 429 status codes from the API. Basically, when rate limit is reached, the processor waits a bit before sending out new payloads.

victor-nitu committed:
[#279] Processor to create (new) PyBossa tasks for new or modified indications

* PyBossa project ID is an environment variable
* New gitignore section.
* [misc] Fix typo in `contrib` processor

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,127 @@
+
+PYBOSSA_URL = os.environ['PYBOSSA_URL']
+PYBOSSA_API_KEY = os.environ['PYBOSSA_API_KEY']
+project_id = 4228
```
This should be an env variable as well

victor-nitu replied:
What if we have multiple projects? Should I create something like `PYBOSSA_PROJECT_INDICATIONS` and follow this convention for subsequent ones?

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+    res = requests.get('{}/api/task'.format(PYBOSSA_URL))
+    if int(res.headers['X-RateLimit-Remaining']) < 30:
+        logger.warn('Rate limit reached, will sleep for 5 minutes')
+        time.sleep(300)  # Sleep for 2 minutes
```
The log message has the wrong time

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+    submit_tasks(tasks)
+
+
+def pybossa_rate_limitation(endpoint):
```
Parameter is unused

vitorbaptista replied:
Rename it to something like `sleep_if_reached_rate_limit`

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+
+def create_tasks(rows):
+    tasks = []
+    for key, group in groupby(rows, lambda x: x['id']):
```
When you're not using some variable, name it `_`. In this case, it would be:

``` python
for _, group in groupby(rows, lambda x: x['id']):
```

Also, if possible, try naming the variables for what they contain. For example, instead of `group`, use `documents` or similar (if they have documents)

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+            task['documents'].append({
+                'name': doc['name'],
+                'url': doc['url']
+            })
```
You can change this for:

``` python
task = {
  'supplement_number': str(doc[0]['supplement_number']),
  # ...
}

task['documents'] = [{'name': doc['name'], 'url': doc['url']}
                     for doc in group]

tasks.append(task)
```

Also, instead of `doc['notes'] or ""` use `doc.get('notes', '')`

Do you need to cast everything to `str()`? Just asking because all fields except `supplement_number` and maybe `id` should be strings already.

victor-nitu replied:
I had issues with those two while working on some other aspects, so I did it for everything. Think it's safe to cast just those two?

As for the list comprehension, yes, that looks exactly like the idea I didn't have. Thanks! 👍

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+    existing_ids = []
+    tasks = []
+    offset = 0
+    limit = 100
```
Not sure how big are the tasks, but 100 seems like a very small limit.

victor-nitu replied:
That's the maximum payload we can get, I'm afraid...

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+            tasks.extend(collection)
+            offset += limit
+        else:
+            break
```
This could be rewritten as:

``` python
while True:
    pybossa_rate_limitation('task')
    last_id = tasks[-1] and tasks[-1].id
    collection = pbc.get_tasks(PROJECT_ID, limit=limit, last_id=last_id)
    tasks.extend(collection)
    if not collection:
        break
```

This removes the need for an offset (which should make PyBossa's query faster, see https://github.com/PyBossa/pybossa-client#on-queries-and-performance), moves the rate limits to inside the `while` loop, and reorganizes the code a bit. I'm not sure if `tasks[-1].id` exists, but there should be a similar call if not.

victor-nitu replied:
Every time we call `pybossa_rate_limitation()` we make an additional request. That's why I tried to reduce its usage.
As for `last_id`, I think `pybossa-client` has it built in. Will check.

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+            break
+    for task in tasks:
+        existing_ids.append(task.info['meta_id'])
+    return existing_ids
```
`return [task.info['meta_id'] for task in tasks]`

victor-nitu replied:
Ah nice, I thought this goes against readability so I didn't do much of it :)

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,126 @@
+
+    # Get existing IDs
+    existing_ids = get_existing_ids(PROJECT_ID)
+    if not isinstance(existing_ids, list):
```
`get_existing_ids` returns either a `list` or throws an exception (if `pbc.get_tasks()` has thrown one), so this check shouldn't be needed. You can use `if existing_ids` if you'd like to know if there're any IDs.

victor-nitu replied:
👍 sounds reasonable

vitorbaptista commented on docker-cloud.yml:
```diff
@@ -253,6 +253,18 @@ pubmed-linker:
+    WAREHOUSE_URL:
+    DATABASE_URL:
+    EXPLORERDB_URL:
+    LOGGING_URL:
```
You also need `PYBOSSA_URL` and others.

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,118 @@
+
+PYBOSSA_URL = os.environ['PYBOSSA_URL']
+PYBOSSA_API_KEY = os.environ['PYBOSSA_API_KEY']
+PROJECT_ID = os.environ['PYBOSSA_PROJECT_INDICATIONS']
```
Instead of hardcoding them here, add those on https://github.com/opentrials/processors/blob/master/processors/base/config.py and get this from the `conf` dict passed to `process()`

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,118 @@
+            'approval_type': first['type'],
+            'approval_id': str(first['id']),
+            'meta_id': str(first['meta_id']),
+            'notes': first.get('notes', '')
```
Do we use things like `supplement_number`, `approval_type` and others in the interface? If not, I would rather not add them here (and don't get them from the SQL query). As far as I can see, the `fda_approval_id` should be enough.

Also, if you're using a specific ID (e.g. `fda_approval_id`), name it as such and not a general `meta_id`, as it serves as a form of documentation on what ID it's about.

victor-nitu replied:
Well, I thought you had something in mind with them when writing the initial query, so I left them there and forgot to ask you back about their importance. Great that you spotted them.

As for the `meta_id`, I fully agree. I sincerely can't remember when and why I came up with it.

Will update in minutes.

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,107 @@
+    for key, group in groupby(rows, lambda x: x['fda_approval_id']):
+        group = list(group)
+        first = group[0]
+        print(str(first['fda_approval_id']))
```
Use `logger` instead of `print` (if this message is needed)

victor-nitu replied:
It isn't needed. It escaped, thanks for spotting it.

victor-nitu replied:
Done.

vitorbaptista commented on processors/base/config.py:
```diff
@@ -47,6 +47,12 @@
+
+PYBOSSA_URL = os.environ.get('PYBOSSA_URL')
+PYBOSSA_API_KEY = os.environ.get('PYBOSSA_API_KEY')
+PROJECT_ID = os.environ.get('PYBOSSA_PROJECT_INDICATIONS')
```
Change to `PYBOSSA_PROJECT_INDICATIONS` (and not a general `PROJECT_ID`)

victor-nitu replied:
I totally agree. Funny though, I actually thought I did it.

victor-nitu replied:
Done.

victor-nitu commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,107 @@
+    FROM documents
+    INNER JOIN fda_approvals ON documents.fda_approval_id = fda_approvals.id
+    INNER JOIN files ON documents.file_id = files.id
+    WHERE documents.url IS NOT NULL
```
Also this has to be changed (missed it).

victor-nitu replied:
Done.

vitorbaptista commented on processors/pybossa_indications/processor.py:
```diff
@@ -0,0 +1,106 @@
+    GROUP BY fda_approvals.id,
+            documents.id,
+            files.id
+    ORDER BY fda_approvals.id
```
Indent the string

vitorbaptista replied:
Don't worry with it, just did it myself to avoid you having to context switch.

vitorbaptista committed:
[#279] Avoid doing requests to check rate limits and general refactorings

opentrials/opentrials#279

vitorbaptista committed:
[#279] Extract PyBossa logic into helper

This will help us when we start having more PyBossa projects.

opentrials/opentrials#279

vitorbaptista commented:
@nightsh Your code was looking fine to me. However, I started looking deeper into it to create the new PyBossa processor for our deduplication project and got a bit carried away. In hindsight, I should've merged your code and then started working on it, but haven't.

I'm assigning it back to you so you can review my changes. I changed the way we check for the rate limits to use only the status code of the response we get (i.e. not doing another request to check the limits). The biggest change was extracting the PyBossa-related code to a separate class (to be reused by our other processors) and adding the logic to keep the tasks in our database in sync with the tasks in PyBossa, adding/removing/updating them as needed.

victor-nitu commented on processors/base/helpers/pybossa_tasks_updater.py:
```diff
@@ -0,0 +1,113 @@
+
+        tasks_to_add = [task_data for task_data in tasks_data
+                        if get_key(task_data) not in existing_ids]
+        tasks_to_delete = [existing_tasks[i]
```
That's an inspired addition to the processor. Good thinking!

victor-nitu commented on processors/base/helpers/pybossa_tasks_updater.py:
```diff
@@ -0,0 +1,113 @@
+                           for i, task_info in enumerate(existing_tasks_info)
+                           if get_key(task_info) not in new_ids]
+
+        tasks_to_update = []
```
Are we allowed to update existing tasks? Even if they already have received answers?

vitorbaptista replied:
Good question. I haven't tested with tasks with answers, but I think the correct behaviour here would be to remove the existing answer and redo the task from the beginning. Hopefully PyBossa does it :pray:

victor-nitu commented:
@vitorbaptista it's a very good addition that further shapes up the desired modularity for our PyBossa related processors. Following this, I would probably rewrite a bit of the indications processor, to make it use more of the logic in the tasks updater, but this will probably be a good candidate for either the hack day or a later update.

Bottom line, I'm quite happy with it. TBH, I didn't run it locally, but by going through all the code I'd say it's merge material, feel free to do it 👍
